nvtabular/io/dataset.py

```diff
@@ -1097,18 +1101,33 @@ def infer_schema(self, n=1):
         Args:
             n (int, optional): Number of rows to sample to infer the dtypes. Defaults to 1.
         """
+
         dtypes = {}
         try:
-            sampled_dtypes = self.sample_dtypes(n)
-            dtypes = dict(zip(sampled_dtypes.index, sampled_dtypes))
+            _ddf = self.to_ddf()
+            dtypes = {
+                col_name: {"dtype": dtype, "is_list": False}
+                for col_name, dtype in _ddf.dtypes.items()
+            }
+            for partition_index in range(_ddf.npartitions):
+                _head = _ddf.partitions[partition_index].head(n)
+
+                if len(_head):
+                    for col in _head.columns:
+                        dtypes[col] = {
+                            "dtype": dispatch._list_val_dtype(_head[col]) or _head[col].dtype,
+                            "is_list": dispatch._is_list_dtype(_head[col]),
+                        }
```

Comment:
I added a special optimization in #1119 to avoid loading any of the partitions from remote storage (which is super slow). It looks like this change will now skip that optimization.  I'll need to double check that this introduces a performance regression in the criteo benchmark on GPC. If it does, I suggest we fix this asap.

Reply:
**Update**: I addressed the previous comment in #1119. While making that change, I also realized that the highlighted code above will always read every partition in the dataset (which I am assuming is due to a missing `break` statement).